File: CSharpAthletic/NetworkProto/ProtoUtils.cs
namespace NetworkProto
{
    // RESPONSES
    public static class ProtoUtils
    {
        public static Response OkResponse()
        {
            return new Response { Type = Response.Types.Type.Ok };
        }

        public static Response ErrorResponse(string error)
        {
            return new Response { Type = Response.Types.Type.Error, Error = error };
        }

        public static Response GetChildrenNoEventsResponse(Domain.DTO.ChildNoEventsDTO[] cnes)
        {
            Response response = new Response { Type = Response.Types.Type.ChildrenEvents };
            foreach(Domain.DTO.ChildNoEventsDTO cne in cnes)
            {
                response.Cne.Add(new ChildNoEventsDTO { Child = GetProtobufChild(cne.Child), NoEvents = cne.NoEvents });
            }
            return response;
        }

        public static Response GetEventsNumberResponse(Domain.DTO.EventCountDTO[] ecs, bool refresh)
        {
            Response response = null;
            if (refresh)
                response = new Response { Type = Response.Types.Type.RefreshEvents };
            else
                response = new Response { Type = Response.Types.Type.EventCount };
            foreach (Domain.DTO.EventCountDTO ec in ecs)
            {
                response.Ec.Add(new EventCountDTO { Event = GetProtobufEvent(ec.Event), AgeGroup = GetProtobufAgeGroup(ec.AgeGroup), Count = ec.Count });
            }
            return response;

        }

        // PROTOBUF REQUEST TO DOMAIN
        public static Domain.User GetUser(Request request)
        {
            return new Domain.User { Username = request.User.Username, Password = request.User.Password };
        }

        public static Domain.DTO.EventAgeGroupDTO GetEventAgeGroupDTO(Request request)
        {
            return new Domain.DTO.EventAgeGroupDTO { Event = GetProtobufEvent(request.Eag.Event), AgeGroup = GetProtobufAgeGroup(request.Eag.AgeGroup) };
        }

        public static Domain.DTO.ParticipantDetailsDTO GetParticipantDetailsDTO(Request request)
        {
            return new Domain.DTO.ParticipantDetailsDTO { Name = request.Pd.Name, Age = request.Pd.Age, Event1 = GetProtobufEvent(request.Pd.Event1), Event2 = GetProtobufEvent(request.Pd.Event2) };
        }

        // DOMAIN TO PROTOBUF
        private static Child GetProtobufChild(Domain.Child child)
        {
            return new Child { Name = child.Name, Age = child.Age };
        }

        private static Event GetProtobufEvent(Domain.Event eventt)
        {
            if(eventt == Domain.Event.NONE)
                return Event.None;
            if (eventt == Domain.Event.E50M)
                return Event.E50M;
            if (eventt == Domain.Event.E100M)
                return Event.E100M;
            if (eventt == Domain.Event.E1000M)
                return Event.E1000M;
            if (eventt == Domain.Event.E1500M)
                return Event.E1500M;
            return Event.None;
        }

        private static AgeGroup GetProtobufAgeGroup(Domain.AgeGroup ageGroup)
        {
            if (ageGroup == Domain.AgeGroup.A6_8Y)
                return AgeGroup.A68Y;
            if (ageGroup == Domain.AgeGroup.A9_11Y)
                return AgeGroup.A911Y;
            if (ageGroup == Domain.AgeGroup.A12_15Y)
                return AgeGroup.A1215Y;
            return AgeGroup.A68Y;
        }

        // PROTOBUF TO DOMAIN
        private static Domain.Child GetProtobufChild(Child child)
        {
            return new Domain.Child { Name = child.Name, Age = child.Age };
        }

        private static Domain.Event GetProtobufEvent(Event eventt)
        {
            if (eventt == Event.None)
                return Domain.Event.NONE;
            if (eventt == Event.E50M)
                return Domain.Event.E50M;
            if (eventt == Event.E100M)
                return Domain.Event.E100M;
            if (eventt == Event.E1000M)
                return Domain.Event.E1000M;
            if (eventt == Event.E1500M)
                return Domain.Event.E1500M;
            return Domain.Event.NONE;
        }

        private static Domain.AgeGroup GetProtobufAgeGroup(AgeGroup ageGroup)
        {
            if (ageGroup == AgeGroup.A68Y)
                return Domain.AgeGroup.A6_8Y;
            if (ageGroup == AgeGroup.A911Y)
                return Domain.AgeGroup.A9_11Y;
            if (ageGroup == AgeGroup.A1215Y)
                return Domain.AgeGroup.A12_15Y;
            return Domain.AgeGroup.A6_8Y;
        }
    }
}

File: JavaAthletic/AthleticNetwork/src/main/java/ProtoUtils.java
import domain.AgeGroup;
import domain.Child;
import domain.Event;
import domain.User;
import dto.ParticipantDetailsDTO;
import dtos.ChildNoEventsDTO;
import dtos.EventAgeGroupDTO;
import dtos.EventCountDTO;

public class ProtoUtils {

    // REQUESTS
    public static Protobuf.Request loginRequest(User user) {
        Protobuf.User userDTO = Protobuf.User.newBuilder().setUsername(user.getUsername()).setPassword(user.getPassword()).build();
        return Protobuf.Request.newBuilder().setType(Protobuf.Request.Type.LOGIN).setUser(userDTO).build();
    }

    public static Protobuf.Request logoutRequest(User user) {
        Protobuf.User userDTO = Protobuf.User.newBuilder()
                .setUsername(user.getUsername())
                .setPassword(user.getPassword()).build();
        return Protobuf.Request.newBuilder().setType(Protobuf.Request.Type.LOGOUT).setUser(userDTO).build();
    }

    public static Protobuf.Request getEventsNumberRequest() {
        return Protobuf.Request.newBuilder().setType(Protobuf.Request.Type.EVENT_COUNT).build();
    }

    public static Protobuf.Request getChildrenNoEventsRequest(EventAgeGroupDTO eag) {
        Protobuf.EventAgeGroupDTO eagDTO = Protobuf.EventAgeGroupDTO.newBuilder()
                .setEvent(getProtobufEvent(eag.getEvent()))
                .setAgeGroup(getProtobufAgeGroup(eag.getAgeGroup())).build();
        return Protobuf.Request.newBuilder().setType(Protobuf.Request.Type.CHILDREN_EVENTS).setEag(eagDTO).build();
    }

    public static Protobuf.Request addParticipantRequest(ParticipantDetailsDTO pd) {
        Protobuf.ParticipantDetailsDTO pdDTO = Protobuf.ParticipantDetailsDTO.newBuilder()
                .setName(pd.getName())
                .setAge(pd.getAge())
                .setEvent1(getProtobufEvent(pd.getEvent1()))
                .setEvent2(getProtobufEvent(pd.getEvent2())).build();
        return Protobuf.Request.newBuilder().setType(Protobuf.Request.Type.ADD_PARTICIPANT).setPd(pdDTO).build();
    }

    // RESPONSES
    public static Protobuf.Response okResponse() {
        return Protobuf.Response.newBuilder().setType(Protobuf.Response.Type.OK).build();
    }

    public static Protobuf.Response errorResponse(String error) {
        return Protobuf.Response.newBuilder().setType(Protobuf.Response.Type.ERROR)
                .setError(error).build();
    }

    public static Protobuf.Response getChildrenNoEventsResponse(ChildNoEventsDTO[] cnes) {
        Protobuf.Response.Builder response = Protobuf.Response.newBuilder().setType(Protobuf.Response.Type.CHILDREN_EVENTS);
        for (ChildNoEventsDTO cne : cnes) {
            Protobuf.ChildNoEventsDTO cneDTO = Protobuf.ChildNoEventsDTO.newBuilder()
                    .setChild(getProtobufChild(cne.getChild()))
                    .setNoEvents(cne.getNoEvents()).build();
            response.addCne(cneDTO);
        }
        return response.build();
    }

    public static Protobuf.Response getEventsNumberResponse(EventCountDTO[] ecs, boolean refresh) {
        Protobuf.Response.Builder response;
        if (refresh)
            response = Protobuf.Response.newBuilder().setType(Protobuf.Response.Type.REFRESH_EVENTS);
        else
            response = Protobuf.Response.newBuilder().setType(Protobuf.Response.Type.EVENT_COUNT);
        for (EventCountDTO ec : ecs) {
            Protobuf.EventCountDTO ecDTO = Protobuf.EventCountDTO.newBuilder()
                    .setEvent(getProtobufEvent(ec.getEvent()))
                    .setAgeGroup(getProtobufAgeGroup(ec.getAgeGroup()))
                    .setCount(ec.getCount()).build();
            response.addEc(ecDTO);
        }
        return response.build();
    }

    // PROTOBUF RESPONSE TO DOMAIN
    public static String getError(Protobuf.Response response) {
        return response.getError();
    }

    public static EventCountDTO[] getEventsNumber(Protobuf.Response response) {
        EventCountDTO[] lst = new EventCountDTO[response.getEcCount()];
        for (int i = 0; i < response.getEcCount(); i++) {
            Protobuf.EventCountDTO ecDTO = response.getEc(i);
            EventCountDTO ec = new EventCountDTO(getProtobufEvent(ecDTO.getEvent()), getProtobufAgeGroup(ecDTO.getAgeGroup()), ecDTO.getCount());
            lst[i] = ec;
        }
        return lst;
    }

    public static ChildNoEventsDTO[] getChildrenNoEvents(Protobuf.Response response) {
        ChildNoEventsDTO[] lst = new ChildNoEventsDTO[response.getCneCount()];
        for (int i = 0; i < response.getCneCount(); i++) {
            Protobuf.ChildNoEventsDTO cneDTO = response.getCne(i);
            ChildNoEventsDTO cne = new ChildNoEventsDTO(getProtobufChild(cneDTO.getChild()), cneDTO.getNoEvents());
            lst[i] = cne;
        }
        return lst;
    }

    // PROTOBUF REQUEST TO DOMAIN
    public static User getUser(Protobuf.Request request) {
        return new User(request.getUser().getUsername(), request.getUser().getPassword());
    }

    public static EventAgeGroupDTO getEventAgeGroupDTO(Protobuf.Request request) {
        return new EventAgeGroupDTO(getProtobufEvent(request.getEag().getEvent()), getProtobufAgeGroup(request.getEag().getAgeGroup()));
    }

    public static ParticipantDetailsDTO getParticipantDetailsDTO(Protobuf.Request request) {
        return new ParticipantDetailsDTO(request.getPd().getName(), request.getPd().getAge(), getProtobufEvent(request.getPd().getEvent1()), getProtobufEvent(request.getPd().getEvent2()));
    }

    // DOMAIN TO PROTOBUF
    private static Protobuf.AgeGroup getProtobufAgeGroup(AgeGroup ageGroup) {
        return switch (ageGroup) {
            case A6_8Y -> Protobuf.AgeGroup.A6_8Y;
            case A9_11Y -> Protobuf.AgeGroup.A9_11Y;
            case A12_15Y -> Protobuf.AgeGroup.A12_15Y;
        };
    }

    private static Protobuf.Event getProtobufEvent(Event event) {
        return switch (event) {
            case NONE -> Protobuf.Event.NONE;
            case E50M -> Protobuf.Event.E50M;
            case E100M -> Protobuf.Event.E100M;
            case E1000M -> Protobuf.Event.E1000M;
            case E1500M -> Protobuf.Event.E1500M;
        };
    }

    private static Protobuf.Child getProtobufChild(Child child) {
        return Protobuf.Child.newBuilder().setName(child.getName()).setAge(child.getAge()).build();
    }

    // PROTOBUF TO DOMAIN
    private static AgeGroup getProtobufAgeGroup(Protobuf.AgeGroup ageGroup) {
        return switch (ageGroup) {
            case A6_8Y -> AgeGroup.A6_8Y;
            case A9_11Y -> AgeGroup.A9_11Y;
            case A12_15Y -> AgeGroup.A12_15Y;
            case UNRECOGNIZED -> null;
        };
    }

    private static Event getProtobufEvent(Protobuf.Event event) {
        return switch (event) {
            case NONE -> Event.NONE;
            case E50M -> Event.E50M;
            case E100M -> Event.E100M;
            case E1000M -> Event.E1000M;
            case E1500M -> Event.E1500M;
            case UNRECOGNIZED -> null;
        };
    }

    private static Child getProtobufChild(Protobuf.Child child) {
        return new Child(child.getName(), child.getAge());
    }

}
